### PR TITLE
feat: improve message content chunks handling

### PR DESCRIPTION
**Pull request by drbh**
This PR improves the parsing of content in messages by deserializing all content into `ContentChunks` (a vector of `ContentChunk`). Each `ContentChunk` is an enum representing either raw text or an image URL. For minimal change impact, ContentChunks can be serialized directly into a flattened string.

### CODE REVIEW

**Comment by drbh** [comment]
**note this PR improves how images are handle when sent as separate content in the chat endpoint. If `` markdown images are included they are still handled the same way, and a future RP may deprecate markdown images in chat and improve how images are sent to the model



example request

```python
from huggingface_hub import InferenceClient

client = InferenceClient("http://127.0.0.1:3000")

chat = client.chat_completion(
    messages=[
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "Whats in this image?"},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/rabbit.png"
                    },
                },
            ],
        },
    ],
    seed=42,
    max_tokens=100,
)
```

File: router/src/infer.rs [hunk 1]
```diff
@@ -362,16 +362,21 @@ impl ChatTemplate {
...
+                            "{:?}\n---\n{}\n{}",
+                            last_message_content, tool_prompt, tools
+                        )
+                        .into(),
+                    );
                 }
             }
         }
 
+        println!("{:?}", messages);
```
Comment (Narsil) [inline router/src/infer.rs:378]: Oops.

File: router/src/lib.rs [hunk 2]
```diff
@@ -896,52 +896,99 @@ pub(crate) struct ImageUrl {
...
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let formatted = self
+            .0
+            .iter()
+            .map(|chunk| match chunk {
+                ContentChunk::Text(s) => s.clone(),
+                ContentChunk::ImageUrl(s) => format!("![]({})", s),
+            })
+            .collect::<Vec<_>>()
+            .join(" ");
```
Comment (Narsil) [inline router/src/lib.rs:950]: why a space ? Shouldn't it be `''`?

File: router/src/lib.rs [hunk 3]
```diff
@@ -896,52 +896,99 @@ pub(crate) struct ImageUrl {
...
+                            .get("text")
+                            .and_then(|t| t.as_str())
+                            .map(|text| Ok(ContentChunk::Text(text.to_string())))
+                            .map_or_else(
+                                || Err(Error::custom("Expected a string or an object")),
+                                |x| x,
+                            ),
+                    },
+                    _ => Err(Error::custom("Expected a string or an object")),
+                })
```
Comment (Narsil) [inline router/src/lib.rs:987]: You shouldn't need this either. Serde should be able to take care of it.

```rust
struct ImageUrl{
    image_url: String
}
```

**Comment by Narsil** [comment]
Can you add some tests too for the deserialization ?

File: router/src/infer.rs (hunk 1 above)
Comment (drbh) [reply router/src/infer.rs]: removed

File: router/src/lib.rs (hunk 2 above)
Comment (drbh) [reply router/src/lib.rs]: oops, removed

File: router/src/lib.rs (hunk 3 above)
Comment (drbh) [reply router/src/lib.rs]: updated and relies on serde to extract the values for both image and text now

**Comment by drbh** [comment]
updates: 

this pr now also parses markdown images into typed `ContentChunks`

for example both the string and structured JSON inputs are deserialized into `ContentChunks` and both currently serialize back into a string

https://github.com/huggingface/text-generation-inference/blob/f8be8d5da7d929d1ce0d7144b0f475a1e99dfcf3/router/src/lib.rs#L1277-L1288

and 

https://github.com/huggingface/text-generation-inference/blob/f8be8d5da7d929d1ce0d7144b0f475a1e99dfcf3/router/src/lib.rs#L1300-L1319

File: router/src/lib.rs [hunk 4]
```diff
@@ -896,52 +897,132 @@ pub(crate) struct ImageUrl {
...
+            })
+            .collect::<Vec<_>>()
+            .join("");
+        serializer.serialize_str(&formatted)
+    }
+}
+
+fn parse_markdown_to_chunks(s: &str) -> Result<Vec<ContentChunk>, serde_json::Error> {
+    let mut chunks = Vec::new();
+    let re = Regex::new(r"!\[([^\]]*)\]\(([^)]+)\)").unwrap();
```
Comment (Narsil) [inline router/src/lib.rs:976]: This is exactly what we *do not* want.

**Comment by drbh** [comment]
closing in favor of https://github.com/huggingface/text-generation-inference/pull/1906

**Pull request closed without merging**